File: solutions/lua/grade-school/1/grade-school.lua
local School = {}

local roster = {}

function School:new()
  setmetatable({}, self)
  roster = {}
  return self
end

function School:roster()
  return roster
end

function School:add(name, grade)
  if roster[grade] == nil then
    roster[grade] = { name }
  else
    table.insert(roster[grade], name)
  end

  table.sort(roster[grade])
end

function School:grade(grade)
  if roster[grade] == nil then return {} end
  return roster[grade]
end

return School

File: solutions/ruby/grade-school/1/grade_school.rb
class School
  attr_reader :roster

  def initialize
    @roster = []
  end

  def students(grade)
    roster
      .select { |s| s.grade == grade }
      .map    { |s| s.name }
      .sort
  end

  def add(name, grade)
    roster << Student.new(name, grade)
  end

  def students_by_grade
    roster
      .map(&:grade).uniq.sort
      .map do |g|
        { grade: g, students: students(g) }
      end
  end

  private

  Student = Struct.new(:name, :grade)
end
